src/wasm-type.h

```diff
@@ -203,25 +225,212 @@ struct Signature {
   }
   bool operator!=(const Signature& other) const { return !(*this == other); }
   bool operator<(const Signature& other) const;
+  std::string toString() const;
 };
 
-std::ostream& operator<<(std::ostream& os, Type t);
-std::ostream& operator<<(std::ostream& os, ParamType t);
-std::ostream& operator<<(std::ostream& os, ResultType t);
-std::ostream& operator<<(std::ostream& os, Signature t);
+struct Field {
+  Type type;
+  enum PackedType : uint32_t {
+    not_packed,
+    i8,
+    i16,
+  } packedType; // applicable iff type=i32
+  bool mutable_;
+
+  Field(Type type, bool mutable_ = false)
+    : type(type), packedType(not_packed), mutable_(mutable_) {}
+  Field(PackedType packedType, bool mutable_ = false)
+    : type(Type::i32), packedType(packedType), mutable_(mutable_) {}
+
+  constexpr bool isPacked() const {
+    if (packedType != not_packed) {
+      assert(type == Type::i32 && "unexpected type");
+      return true;
+    }
+    return false;
+  }
+
+  bool operator==(const Field& other) const {
+    return type == other.type && packedType == other.packedType &&
+           mutable_ == other.mutable_;
+  }
+  bool operator!=(const Field& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+typedef std::vector<Field> FieldList;
+
+struct Struct {
+  FieldList fields;
```

Comment:
Hmm, there are really two layers here. A `Struct` for example can end up in the type section just like a `Signature`, but entries in the type section are not nullable but just describe shapes, like `Signature` always was. A type referencing such a shape, e.g. in the form of a `ref struct ...` or `ref func ...` is what can be nullable, which I tried to model by pulling the nullable property into the specific typedef right where the concept of something being nullable begins, with `Tuple` being special in that it can't be nullable. If we pull the nullable property further out, we'll get a type section potentially containing the same `Signature` for example twice, once nullable and once non-nullable, which is wrong. Ofc we can disallow adding nullable signatures there, but then there's still that unused `nullable` property where stuff can go wrong.

Reply:
I see, thanks. I think this makes sense the way it is, then. I would still drop the extra wrapper around `Tuple`, though.